src/URI.cpp

```diff
@@ -0,0 +1,124 @@
+#include "URI.h"
+#include "logger.h"
+#include <iostream>
+
+namespace uri {
+
+using std::array;
+using std::move;
+
+static std::string findTopicFromPath(std::string URIString) {
+  auto Path = URIString.find("/");
+  if (Path == 0) {
+    URIString = URIString.substr(1);
+  }
+  Path = URIString.find("/");
+  if (Path == std::string::npos) {
+    return URIString;
+  } else {
+    if (Path == 0) {
+      return URIString.substr(1);
+    } else {
+      return std::string();
+    }
+  }
+}
+
+void URI::updateHostPortAndTopic() {
+  if (Port != 0) {
+    HostPort = fmt::format("{}:{}", Host, Port);
+  } else {
+    HostPort = Host;
+  }
+  auto TopicFromPath = findTopicFromPath(Path);
+  if (not TopicFromPath.empty()) {
+    Topic = TopicFromPath;
+  }
+}
+
+URI::URI(std::string URIString) { parse(URIString); }
+
+static vector<std::string> findProtocol(const std::string &URIString) {
+  auto Slashes = URIString.find("://");
+  if (Slashes == std::string::npos or Slashes == 0) {
+    return {std::string(), URIString};
+  }
+  auto Protocol = URIString.substr(0, Slashes);
+  if (Protocol.find_first_of("0123456789") != std::string::npos) {
```

Comment:
This seems a bit brittle to me. What happens if someone accidentally does `URI.parse("http:://my.Host")`?
BTW: there should be a unit test for this.

I think using a simple regex would be a better option.